---
title: Terminology and Data Acquisition
weight: 10
chapter: false
pre: "<b>3.1 </b>"
---

## Data Terminology

### What is a read?
A read is a string of bases represented by their one letter codes. Here is an example of a read that is 100 bases long. `TTAACCTTGGTTTTGAACTTGAACACTTAGGGGATTGAAGATTCAACAACCCTAAAGCTTGGGGTAAAACCTTAGGGGAT`

### What is a contig?
A contig is the consensus sequence generated by aligning reads to themselves.

{{< figure src="https://isugenomics.github.io/bioinformatics-workbook/introduction/assets/contig.png" title="Alignment and Consensus" >}}
	
The last line is the consensus of the aligned reads which we call a contig.

### What is a scaffold?
A scaffold is a set of contigs that have been ordered and oriented based on mate pair or long distance information. Usually there are large gaps found between contigs. New sequencing technologies are helping to close those gaps. 

### What is a chromosome?

Chromosomes are the largest DNA molecules in the cell. Scaffolds can be ordered and oriented using a genetic map or HiC data into linkage groups or chromosomes. The ultimate goal of a genome assembly project is to assemble reads into chromosomes.

### Next Generation Sequencing
Next-generation sequencing generates masses of DNA sequence data that's richer and more complete than is imaginable with Sanger sequencing. Illumina sequencing systems can deliver data output ranging from 300 kilobases up to multiple terabases in a single run, depending on instrument type and configuration.

### Sequence Assembly
Sequence assembly refers to aligning and merging fragments from a longer DNA sequence in order to reconstruct the original sequence. The assembly can be found at different stages of completion. It could be conting, scaffold or even chromosome level.

### Sequence Annotation
DNA annotation or genome annotation is the process of identifying the locations of genes and all of the coding regions in a genome and determining what those genes do.
{{< figure src="/images/genome_ann.jpg" title="Genes and Genome" >}}

## Data Acquisition

{{% notice note %}}
While using coding and bioinformatics, we want to avoid using graphical user interphases (GUIs) as much as we can. Can somebody answer why?

{{% /notice %}}

 
 There are exceptions, where we already know the data structure of a file. However, visualizing large datasets as the ones that we are going to download, could be practically impossible since we need to open the whole document to process it.
 
 [NCBI](https://www.ncbi.nlm.nih.gov/) which stands for the The National Center for Biotechnology Information helps bioinformaticians by providing access to biomedical and genomic information. This is a GUI which helps people search and process data. This is given the fact that the data has been previously curated by others.
 
 They also have access to an [FTP site](ftp://ftp.ncbi.nlm.nih.gov), where we can download files in bulk. Usually, exploring these folders become difficult, therefore, they create summary files that we can play with. 
 
 Let's start by downloading the `assembly_summary_genbank.txt` from the NCBI FTP site at ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/. If you click on this site link the directory content should look something like this:
 
	.
	├── [  30]  README_assembly_summary.txt -> ../README_assembly_summary.txt
	├── [  27]  README_change_notice.txt -> ../README_change_notice.txt
	├── [ 22K]  species_genome_size.txt.gz
	├── [ 31M]  ANI_report_bacteria.txt
	├── [2.2M]  assembly_summary_genbank_historical.txt
	├── [ 46M]  assembly_summary_genbank.txt
	├── [2.4M]  assembly_summary_refseq_historical.txt
	└── [ 34M]  assembly_summary_refseq.txt

### Back to the terminal	
Create a new folder in your `/scratch/username` directory named `afternoon` and move inside the `afternoon` folder:

```bash
$ mkdir afternoon
$ cd afternoon
```

If you type `pwd` you whould get something like this:
```bash
$ pwd
```
	/scratch/asecas86/afternoon
	
	
Now, let us retrieve the `assembly_summary_genbank.txt` by typing the following code in your terminal

```bash
$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt
```

		--2018-03-04 22:46:48--  ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt
				   => “assembly_summary_genbank.txt”
		Resolving ftp.ncbi.nlm.nih.gov... 165.112.9.229, 2607:f220:41e:250::7
		Connecting to ftp.ncbi.nlm.nih.gov|165.112.9.229|:21... connected.
		Logging in as anonymous ... Logged in!
		==> SYST ... done.    ==> PWD ... done.
		==> TYPE I ... done.  ==> CWD (1) /genomes/ASSEMBLY_REPORTS ... done.
		==> SIZE assembly_summary_genbank.txt ... 48113668
		==> PASV ... done.    ==> RETR assembly_summary_genbank.txt ... done.
		Length: 48113668 (46M) (unauthoritative)

		100%[=====================================================================================================================================>] 48,113,668  11.1M/s   in 8.4s

Check for the downloaded file 
```bash
$ ls -lth
```

	total 52M
	-rw------- 1 asecas86 clusterusers 46M Mar  4 22:46 assembly_summary_genbank.txt

