---
title: Terminology and Data Acquisition
weight: 10
chapter: false
pre: "<b>3.1 </b>"
---

## Data Terminology

### What is a read?
A read is a string of bases represented by their one letter codes. Here is an example of a read that is 100 bases long. `TTAACCTTGGTTTTGAACTTGAACACTTAGGGGATTGAAGATTCAACAACCCTAAAGCTTGGGGTAAAACCTTAGGGGAT`

### What is a contig?
A contig is the consensus sequence generated by aligning reads to themselves.

{{< figure src="https://isugenomics.github.io/bioinformatics-workbook/introduction/assets/contig.png" title="Alignment and Consensus" >}}
	
The last line is the consensus of the aligned reads which we call a contig.

### What is a scaffold?
A scaffold is a set of contigs that have been ordered and oriented based on mate pair or long distance information. Usually there are large gaps found between contigs. New sequencing technologies are helping to close those gaps. 

### What is a chromosome?

Chromosomes are the largest DNA molecules in the cell. Scaffolds can be ordered and oriented using a genetic map or HiC data into linkage groups or chromosomes. The ultimate goal of a genome assembly project is to assemble reads into chromosomes.

### Next Generation Sequencing
Next-generation sequencing generates masses of DNA sequence data that's richer and more complete than is imaginable with Sanger sequencing. Illumina sequencing systems can deliver data output ranging from 300 kilobases up to multiple terabases in a single run, depending on instrument type and configuration.

### Sequence Assembly
Sequence assembly refers to aligning and merging fragments from a longer DNA sequence in order to reconstruct the original sequence. The assembly can be found at different stages of completion. It could be conting, scaffold or even chromosome level.

### Sequence Annotation
DNA annotation or genome annotation is the process of identifying the locations of genes and all of the coding regions in a genome and determining what those genes do.
{{< figure src="/images/genome_ann.jpg" title="Genes and Genome" >}}

## Data Acquisition

{{% notice note %}}
While using coding and bioinformatics, we want to avoid using graphical user interphases (GUIs) as much as we can. Can somebody answer why?
{{% /notice %}}

 
Visualizing large datasets (as the ones that we are going to download) could be practically impossible.  Opening the whole document will likely **crash your computer**.
 
 {{< figure src="https://media.giphy.com/media/l3q2Mdm0cRPLPcuGI/giphy.gif" title="" >}}
 
 However, there are exceptions, where we already know the data structure of a file and we can start running codes on it.  
 
 [NCBI](https://www.ncbi.nlm.nih.gov/) which stands for the The National Center for Biotechnology Information helps bioinformaticians by providing access to biomedical and genomic information. This is a GUI which helps people search and process data. This is given the fact that the data has been previously curated by others.
 
 They also have access to an [FTP site](ftp://ftp.ncbi.nlm.nih.gov), where we can download files in bulk. Usually, exploring these folders become difficult, therefore, they create summary files that we can play with. 
 
 Let's start by downloading the `assembly_summary_genbank.txt` from the NCBI FTP site at ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/. If you click on this site link the directory content should look something like this:
 
	.
	├── [  30]  README_assembly_summary.txt -> ../README_assembly_summary.txt
	├── [  27]  README_change_notice.txt -> ../README_change_notice.txt
	├── [ 22K]  species_genome_size.txt.gz
	├── [ 31M]  ANI_report_bacteria.txt
	├── [2.2M]  assembly_summary_genbank_historical.txt
	├── [ 46M]  assembly_summary_genbank.txt
	├── [2.4M]  assembly_summary_refseq_historical.txt
	└── [ 34M]  assembly_summary_refseq.txt

### Back to the terminal	
Create a new folder in your `/scratch/username` directory named `afternoon` and move inside the `afternoon` folder:

```bash
$ mkdir afternoon
$ cd afternoon
```

If you type `pwd` you whould get something like this:
```bash
$ pwd
```
	/scratch/asecas86/afternoon
	
	
Now, let us retrieve the `assembly_summary_genbank.txt` by typing the following code in your terminal

```bash
$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt
```

		--2018-03-04 22:46:48--  ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt
				   => “assembly_summary_genbank.txt”
		Resolving ftp.ncbi.nlm.nih.gov... 165.112.9.229, 2607:f220:41e:250::7
		Connecting to ftp.ncbi.nlm.nih.gov|165.112.9.229|:21... connected.
		Logging in as anonymous ... Logged in!
		==> SYST ... done.    ==> PWD ... done.
		==> TYPE I ... done.  ==> CWD (1) /genomes/ASSEMBLY_REPORTS ... done.
		==> SIZE assembly_summary_genbank.txt ... 48113668
		==> PASV ... done.    ==> RETR assembly_summary_genbank.txt ... done.
		Length: 48113668 (46M) (unauthoritative)

		100%[=====================================================================================================================================>] 48,113,668  11.1M/s   in 8.4s

Check for the downloaded file 
```bash
$ ls -lth
```

	total 52M
	-rw------- 1 asecas86 clusterusers 46M Mar  4 22:46 assembly_summary_genbank.txt

We can explore the file without completely opening it

```bash
$ head assembly_summary_genbank.txt
```

	#   See ftp://ftp.ncbi.nlm.nih.gov/genomes/README_assembly_summary.txt for a description of the columns in this file.
	# assembly_accession    bioproject      biosample       wgs_master      refseq_category taxid   species_taxid   organism_name   infraspecific_name      isolate       version_status  assembly_level  release_type    genome_rep      seq_rel_date    asm_name        submitter       gbrs_paired_asm paired_asm_comp ftp_path      excluded_from_refseq    relation_to_type_material
	GCA_000001215.4 PRJNA13812      SAMN02803731            reference genome        7227    7227    Drosophila melanogaster                 latest  Chromosome   Major    Full    2014/08/01      Release 6 plus ISO1 MT  The FlyBase Consortium/Berkeley Drosophila Genome Project/Celera Genomics       GCF_000001215.4 identical     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/215/GCA_000001215.4_Release_6_plus_ISO1_MT
	GCA_000001405.27        PRJNA31257                      reference genome        9606    9606    Homo sapiens                    latest  Chromosome      PatchFull     2017/12/21      GRCh38.p12      Genome Reference Consortium     na      na      ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.27_GRCh38.p12
	GCA_000001515.5 PRJNA13184      SAMN02981217    AACZ00000000.4  representative genome   9598    9598    Pan troglodytes         Yerkes chimp pedigree #C0471 (Clint)  latest  Chromosome      Major   Full    2016/05/03      Pan_tro 3.0     Chimpanzee Sequencing and Analysis Consortium   GCF_000001515.7 different    ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/515/GCA_000001515.5_Pan_tro_3.0
	GCA_000001545.3 PRJNA20869      SAMN02981238    ABGA00000000.1  representative genome   9601    9601    Pongo abelii            ISIS 71 latest  Chromosome   Major    Full    2008/11/13      P_pygmaeus_2.0.2        Orangutan Genome Sequencing Consortium  na      na      ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/545/GCA_000001545.3_P_pygmaeus_2.0.2
	GCA_000001635.8 PRJNA20689                      reference genome        10090   10090   Mus musculus                    latest  Chromosome      Patch   Full 2017/09/15       GRCm38.p6       Genome Reference Consortium     GCF_000001635.26        identical       ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.8_GRCm38.p6
	GCA_000001735.1 PRJNA10719      SAMN03081427            reference genome        3702    3702    Arabidopsis thaliana    ecotype=Columbia                latestChromosome      Major   Full    2011/05/11      TAIR10  The Arabidopsis Information Resource (TAIR)     GCF_000001735.3 different       ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/735/GCA_000001735.1_TAIR10
	GCA_000001765.2 PRJNA10626      SAMN00779672    AADE00000000.1  representative genome   46245   7237    Drosophila pseudoobscura pseudoobscura  strain=MV2-25latest   Chromosome      Major   Full    2013/04/11      Dpse_3.0        Baylor College of Medicine      GCF_000001765.3 identical       ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/765/GCA_000001765.2_Dpse_3.0
	GCA_000001895.4 PRJNA10629      SAMN02808228    AABR00000000.7  representative genome   10116   10116   Rattus norvegicus       strain=mixed            latestChromosome      Major   Full    2014/07/01      Rnor_6.0        Rat Genome Sequencing Consortium        GCF_000001895.5 identical       ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/895/GCA_000001895.4_Rnor_6.0
	
Can we make sense of the data from here?  Really difficult to tell right??? That's why it is important to **ALWAYS CREATE A README FILE**.  In their data structure above we can see that they have `readme files`. We can download those files.

```bash
$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/README_assembly_summary.txt
```

Usually readme files can be opened with any text editor (nano or vim), in this case we will use `less`

```bash
$ less README_assembly_summary.txt
```

Keep pushing `enter` until you reach the end of the document and tell me what information about their ftp archives can be recovered and are useful. Take notes of what information we can get on each column.

{{% notice warning %}}
Push `Q` to quit watching the `readme` file.
{{% /notice %}}

	################################################################################
	README for the assembly_summary_genbank.txt, assembly_summary_refseq.txt and 
	assembly_summary.txt files found on the NCBI genomes FTP site:
	  ftp://ftp.ncbi.nlm.nih.gov/genomes 

	Last updated: February 25, 2018
	################################################################################

	======================
	ASSEMBLY SUMMARY FILES
	======================

	The assembly_summary files report metadata for the genome assemblies on the 
	NCBI genomes FTP site.

	Four master files reporting data for either GenBank or RefSeq genome assemblies 
	are available under ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/
	assembly_summary_genbank.txt            - current GenBank genome assemblies
	assembly_summary_genbank_historical.txt - replaced and suppressed GenBank genome
											  assemblies
	assembly_summary_refseq.txt             - current RefSeq genome assemblies
	assembly_summary_refseq_historical.txt  - replaced and suppressed RefSeq genome
											  assemblies

	assembly_summary_genbank.txt and assembly_summary_genbank_historical.txt are 
	also available at:
	ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt
	ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank_historical.txt


	assembly_summary_refseq.txt and assembly_summary_refseq_historical.txt are 
	also available at:
	ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt
	ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq_historical.txt

	The assembly_summary.txt files in the directories named for taxonomic groups or
	species contain the relevant subsets of the data from the master files.


	=====================
	COLUMN SPECIFICATIONS
	=====================

	The assembly_summary.txt files have 20 tab-delimited columns. 
	Header rows begin with '#".

	Column  1: "assembly_accession"
	   Assembly accession: the assembly accession.version reported in this field is 
	   a unique identifier for the set of sequences in this particular version of 
	   the genome assembly.
	  
	Column  2: "bioproject"
	   BioProject: accession for the BioProject which produced the sequences in the 
	   genome assembly. A BioProject is a collection of biological data related to a
	   single initiative, originating from a single organization or from a 
	   consortium. A BioProject record provides users a single place to find links 
	   to the diverse data types generated for that project. The record can be 
	   retrieved from the NCBI BioProject resource:
	   https://www.ncbi.nlm.nih.gov/bioproject/
	   
	Column  3: "biosample"
	   BioSample: accession for the BioSample from which the sequences in the genome
	   assembly were obtained. A BioSample record contains a description of the 
	   biological source material used in experimental assays. The record can be 
	   retrieved from the NCBI BioSample resource:
	   https://www.ncbi.nlm.nih.gov/biosample/
	   
	Column  4: "wgs_master"
	   WGS-master: the GenBank Nucleotide accession and version for the master 
	   record of the Whole Genome Shotgun (WGS) project for the genome assembly. The
	   master record can be retrieved from the NCBI Nucleotide resource: 
	   https://www.ncbi.nlm.nih.gov/nuccore
	   Genome assemblies that are complete genomes, and those that are clone-based,
	   do not have WGS-master records in which case this field will be empty.
	   
	Column  5: "refseq_category"
	   RefSeq Category: whether the assembly is a reference or representative genome
	   in the NCBI Reference Sequence (RefSeq) project classification. 
	   Values:
			   reference genome      - a manually selected high quality genome 
									   assembly that NCBI and the community have 
									   identified as being important as a standard 
									   against which other data are compared
			   representative genome - a genome computationally or manually selected
									   as a representative from among the best 
									   genomes available for a species or clade that
									   does not have a designated reference genome
			   na                    - no RefSeq category assigned to this assembly
	   Prokaryotes may have more than one reference or representative genome per 
	   species. For more information see: 
	   https://www.ncbi.nlm.nih.gov/refseq/about/prokaryotes/#referencegenome
	   Eukaryotes have no more than one reference or representative genome per 
	   species. If there are no assemblies in RefSeq for a particular eukaryotic 
	   species, then the GenBank assembly that RefSeq would select as the best 
	   available for that species will be designated as the representative genome.
	   Viruses may have one or more reference genomes per species. The 
	   representative genome designation is not applied to viruses and viroids.

	Column  6: "taxid"
	   Taxonomy ID: the NCBI taxonomy identifier for the organism from which the 
	   genome assembly was derived. The NCBI Taxonomy Database is a curated 
	   classification and nomenclature for all of the organisms in the public 
	   sequence databases. The taxonomy record can be retrieved from the NCBI 
	   Taxonomy resource:
	   https://www.ncbi.nlm.nih.gov/taxonomy/
	   
	Column  7: "species_taxid"
	   Species taxonomy ID: the NCBI taxonomy identifier for the species from which 
	   the genome assembly was derived. The species taxid will differ from the 
	   organism taxid (column 6) only when the organism was reported at a sub-
	   species or strain level.
	   
	Column  8: "organism_name"
	   Organism name: the scientific name of the organism from which the sequences 
	   in the genome assembly were derived. This name is taken from the NCBI 
	   Taxonomy record for the taxid specified in column 6. Some older taxids were 
	   assigned at the strain level and for these the organism name will include the
	   strain. Current practice is only to assign taxids at the species level; for 
	   these the organism name will be just the species, however, the strain name 
	   will be reported in the infraspecific_name field (column 9).

	Column  9: "infraspecific_name"
	   Infraspecific name: the strain, breed, cultivar or ecotype of the organism 
	   from which the sequences in the genome assembly were derived. Data are 
	   reported in the form tag=value, e.g. strain=AF16. Strain, breed, cultivar 
	   and ecotype are not expected to be used together, however, if they are then 
	   they will be reported in a list separated by ", /". Empty if no strain, 
	   breed, cultivar or ecotype is specified on the genomic sequence records.
	  
	Column 10: "isolate"
	   Isolate: the individual isolate from which the sequences in the genome 
	   assembly were derived. Empty if no isolate is specified on the genomic 
	   sequence records.

	Column 11: "version_status"
	   Version status: the release status for the genome assembly version.
	   Values:
			   latest     - the most recent of all the versions for this assembly 
							chain
			   replaced   - this version has been replaced by a newer version of the
							assembly in the same chain
			   suppressed - this version of the assembly has been suppressed 
	   An assembly chain is the collection of all versions for the same assembly 
	   accession.

	Column 12: "assembly_level"
	   Assembly level: the highest level of assembly for any object in the genome 
	   assembly.
	   Values:
		  Complete genome - all chromosomes are gapless and have no runs of 10 or 
							more ambiguous bases (Ns), there are no unplaced or 
							unlocalized scaffolds, and all the expected chromosomes
							are present (i.e. the assembly is not noted as having 
							partial genome representation). Plasmids and organelles
							may or may not be included in the assembly but if 
							present then the sequences are gapless.
		  Chromosome      - there is sequence for one or more chromosomes. This 
							could be a completely sequenced chromosome without gaps
							or a chromosome containing scaffolds or contigs with 
							gaps between them. There may also be unplaced or 
							unlocalized scaffolds.
		  Scaffold        - some sequence contigs have been connected across gaps to
							create scaffolds, but the scaffolds are all unplaced or 
							unlocalized.
		  Contig          - nothing is assembled beyond the level of sequence 
							contigs

	Column 13: "release_type"
	   Release type: whether this version of the genome assembly is a major, minor 
	   or patch release.
	   Values:
			   Major - changes from the previous assembly version result in a 
					   significant change to the coordinate system. The first 
					   version of an assembly is always a major release. Most 
					   subsequent genome assembly updates are also major releases.
			   Minor - changes from the previous assembly version are limited to the
					   following changes, none of which result in a significant 
					   change to the coordinate system of the primary assembly-unit:
					   - adding, removing or changing a non-nuclear assembly-unit
					   - dropping unplaced or unlocalized scaffolds
					   - adding up to 50 unplaced or unlocalized scaffolds which are
						 shorter than the current scaffold-N50 value
					   - replacing a component with a gap of the same length
			   Patch - the only change is the addition or modification of a patch 
					   assembly-unit. 
	   See the NCBI Assembly model web page (https://www.ncbi.nlm.nih.gov/assembly/
	   model/#asmb_def) for definitions of assembly-units and genome patches.

	Column 14: "genome_rep"
	   Genome representation: whether the goal for the assembly was to represent the
	   whole genome or only part of it.
	   Values:
		  Full    - the data used to generate the assembly was obtained from the 
					whole genome, as in Whole Genome Shotgun (WGS) assemblies for 
					example. There may still be gaps in the assembly.
		  Partial - the data used to generate the assembly came from only part of 
					the genome. 
	   Most assemblies have full genome representation with a minority being partial
	   genome representation. See the Assembly help web page 
	   (https://www.ncbi.nlm.nih.gov/assembly/help/) for reasons that the genome 
	   representation would be set to partial.

	Column 15: "seq_rel_date"
	   Sequence release date: the date the sequences in the genome assembly were 
	   released in the International Nucleotide Sequence Database Collaboration 
	   (INSDC) databases, i.e. DDBJ, ENA or GenBank.

	Column 16: "asm_name"
	   Assembly name: the submitter's name for the genome assembly, when one was 
	   provided, otherwise a default name, in the form ASM#####v#, is provided by 
	   NCBI. Assembly names are not unique.

	Column 17: "submitter"
	   Submitter: the submitting consortium or first position if a list of 
	   organizations. The full submitter information is available in the NCBI 
	   BioProject resource: www.ncbi.nlm.nih.gov/bioproject/

	Column 18: "gbrs_paired_asm"
	   GenBank/RefSeq paired assembly: the accession.version of the GenBank assembly
	   that is paired to the given RefSeq assembly, or vice-versa. "na" is reported 
	   if the assembly is unpaired.

	Column 19: "paired_asm_comp"
	   Paired assembly comparison: whether the paired GenBank & RefSeq assemblies 
	   are identical or different.
	   Values:
		  identical - GenBank and RefSeq assemblies are identical
		  different - GenBank and RefSeq assemblies are not identical
		  na        - not applicable since the assembly is unpaired

	Column 20: "ftp_path"
	   FTP path: the path to the directory on the NCBI genomes FTP site from which 
	   data for this genome assembly can be downloaded.

	Column 21: "excluded_from_refseq"
	   Excluded from RefSeq: reasons the assembly was excluded from the NCBI 
	   Reference Sequence (RefSeq) project, including any assembly anomalies. See:
	   https://www.ncbi.nlm.nih.gov/assembly/help/anomnotrefseq/

	Column 22: "relation_to_type_material"
	   Relation to type material: contains a value if the sequences in the genome 
	   assembly were derived from type material.
	   Values:
		  assembly from type material - the sequences in the genome assembly were 
			 derived from type material
		  assembly from synonym type material - the sequences in the genome assembly
			 were derived from synonym type material
		  assembly from proxytype material - the sequences in the genome assembly 
			 were derived from proxytype material
		  assembly designated as neotype - the sequences in the genome assembly 
			 were derived from neotype material
		  assembly from reference material - the sequences in the genome assembly 
			 were derived from reference material where type material never was 
			 available and is not likely to ever be available


	=====================================
	HOW TO USE THE ASSEMBLY SUMMARY FILES
	=====================================

	The metadata provided in the assembly_summary.txt files can be used to identify
	assemblies of interest for subsequent download. 

	The Genomes FTP FAQ provides examples of how to use the assembly_summary.txt 
	files to download sets of assemblies. See: 

	How can I download only the current version of each assembly?
	https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/#current

	How can I download RefSeq data for all complete bacterial genomes?
	https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/#allcomplete

	Other sets of assemblies of interest can be downloaded using variations on 
	these instructions.

	________________________________________________________________________________
	National Center for Biotechnology Information (NCBI)
	National Library of Medicine
	National Institutes of Health
	8600 Rockville Pike
	Bethesda, MD 20894, USA
	tel: (301) 496-2475
	fax: (301) 480-9241
	e-mail: info@ncbi.nlm.nih.gov
	________________________________________________________________________________

Let's focus on our organism of interest `Aspergillus flavus`. Let's try searching for our pathogen in the databases

```bash
$ grep "aspergillus" assembly_summary_genbank.txt
```
What did you get as output from the previous code? 

**NOTHING? HOW COME??**

Try this
```bash
$ grep -i "aspergillus" assembly_summary_genbank.txt
```
{{% notice note %}}
Take 1 minute to understand why this happened
{{% /notice %}}

Let's count how many entries we have and at the same time redirect our entries to a file named `aspergillus_assemblies.txt`

```bash
$ grep -ic "aspergillus" assembly_summary_genbank.txt
$ grep -i "aspergillus" assembly_summary_genbank.txt > aspergillus_assemblies.txt
```
	110

Remember that there can be different types of assembly levels. Which shall we be interested in this case. We want `Scaffold` level assemblies if possible, then less filter our output. For these purposes let me introduce you to `awk`

```bash
awk -F'\t' '$12 == "Chromosome"' aspergillus_assemblies.txt | wc -l

```
		3
		
{{% notice note %}}
**Can somebody tell me what are we doing with this code?**
{{% /notice %}}
```bash
awk -F'\t' '$12 == "Chromosome"' aspergillus_assemblies.txt > chr_aspergillus_assemblies.txt
```
### FIVE Minute task
Let's retrieve the column where the ftp address to download those 3 chromosomal assemblies and write it to a new file. 

{{% notice tip %}}
Use `awk` and select a different column number
{{% /notice %}}