<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
    <meta name="generator" content="Hugo 0.37" />
    <meta name="description" content="Entomology and Plant Pathology Bioinformatics workshop">
<meta name="author" content="Andres Espindola">

    <link rel="shortcut icon" href="/images/favicon.png" type="image/x-icon" />
<link rel="icon" href="/images/favicon.png" type="image/x-icon" />
    <title>Data Acquisition :: Introduction to Programming for beginners</title>
    
    
    <link href="/css/nucleus.css?1521515504" rel="stylesheet">
    <link href="/css/font-awesome.min.css?1521515504" rel="stylesheet">
    <link href="/css/hybrid.css?1521515504" rel="stylesheet">
    <link href="/css/featherlight.min.css?1521515504" rel="stylesheet">
    <link href="/css/perfect-scrollbar.min.css?1521515504" rel="stylesheet">
    <link href="/css/auto-complete.css?1521515504" rel="stylesheet">
    <link href="/css/theme.css?1521515504" rel="stylesheet">
    <link href="/css/hugo-theme.css?1521515504" rel="stylesheet">
    
      <link href="/css/theme-mine.css?1521515504" rel="stylesheet">
    

    <script src="/js/jquery-2.x.min.js?1521515504"></script>
    
    <style type="text/css">
      :root #header + #content > #left > #rlblock_left{ 
          display:none !important;
      }
      
    </style>
    
<style type="text/css">
    body {
  font-family: 'Roboto', sans-serif;
  font-size: 16px;
  color: rgba(0,0,0,0.8);
  line-height: 1.65;
  font-size: 1rem;
		}
		
@media screen and (min-width: 58em) {
  body {
    font-size: 20px;
  }
}

h1 {
    font-family: 'Montserrat', sans-serif;
    font-weight: 400;
    margin-bottom: .5rem;
    line-height: 1.25;
    color: #313131;
    text-rendering: optimizeLegibility;
}

<!-- Making the center more smaller -->
#body .padding {
    padding: 3rem 7rem;
}
</style>


  </head>
  <body class="" data-url="/afternoon/data_acquisition/">
    <nav id="sidebar" class="showVisitedLinks">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo" href="/">

<svg xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns="http://www.w3.org/2000/svg" enable-background="new 0 0 386.912 250.934" xml:space="preserve" height="100%" viewBox="0 0 386.912 250.934" width="100%" version="1.1" y="0px" x="0px" xmlns:cc="http://creativecommons.org/ns#" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:dc="http://purl.org/dc/elements/1.1/">
<g><path d="m281.32 239.69h3.882c2.043 0 2.414 1.507 2.414 2.17 0 0.841-0.426 1.568-1.182 1.826 0.625 0.277 0.969 0.515 0.969 1.946 0 1.122 0 1.4 0.343 1.531v0.206h-1.774c-0.106-0.368-0.172-0.763-0.172-1.552 0-1.051-0.065-1.456-1.253-1.456h-1.654v3.008h-1.572v-7.679zm3.453 3.377c0.831 0 1.243-0.288 1.243-1.051 0-0.404-0.186-1.002-1.126-1.002h-1.998v2.053h1.881z" clip-path="url(#a)"/><path d="m276.94 243.66c0-4.013 3.251-7.277 7.281-7.277 4.013 0 7.271 3.258 7.271 7.277 0 4.02-3.265 7.264-7.271 7.264-4.024 0-7.275-3.251-7.281-7.264zm0.789 0c0 3.588 2.897 6.481 6.492 6.481 3.576-0.007 6.48-2.894 6.48-6.481 0-3.584-2.904-6.488-6.48-6.495-3.595 0.007-6.482 2.911-6.492 6.495z" clip-path="url(#a)"/><path d="m334.75 87.149c3.629-10.06 18.174-50.376 18.174-50.376l-49.423-36.766h-138.41c-2.492-0.05-16.191 0.017-31.845 6.028-20.634 7.924-35.622 22.885-43.349 43.269-4.257 11.25-9 24.002-14.064 37.845h-25.11c-14.031 0-24.918 5.835-28.402 15.231-1.862 5.01-22.32 62.43-22.32 62.43h47.982c-3.512 10.017-17.236 49.315-17.236 49.315l51.547 35.712 138.78-0.011c2.308-0.082 56.655-2.529 70.987-43.597 0 0 11.023-30.923 14.769-41.42h28.712c15.547 0 25.393-5.417 29.261-16.086l22.114-61.576c0 0.001-41.218 0.004-52.165 0.004m-200.18 104.86c2.124-5.907 4.212-11.129 4.212-11.129h69.375s-2.925 7.594-4.191 11.129h-69.396m118.04-135.48-4.216 11.772h-69.209l4.232-11.772h69.193z" clip-path="url(#a)" fill="#bebebe"/><path d="m327.3 92.427c9.461-26.24 16.828-46.647 17.014-47.164l2.372-6.567-44.919-33.411h-136.68c-3.61-0.067-52.86-0.091-70.297 45.919-4.952 13.081-10.154 27.13-15.309 41.223h-28.756c-11.363 0-20.712 4.32-23.496 11.824-1.871 5.034-19.806 55.366-19.806 55.366h47.941c-9.382 26.528-15.902 45.375-16.043 45.773l-2.328 6.742 46.942 32.522h136.78l0.168-0.011c2.159-0.079 53.071-2.354 66.239-40.092 2.691-7.73 8.774-24.758 16.018-44.936h32.402c11.359 0 20.875-3.083 24.339-12.626 3.457-9.55 19.591-54.564 19.591-54.564h-52.171m-67.674-40.379-7.483 20.977h-80.332l7.52-20.98c-0.001 0 61.761 0.003 80.295 0.003m-132.49 145.19 8.328-21.95h80.479l-8.287 21.95h-80.52z" clip-path="url(#a)"/></g><g stroke="#ff7300" fill="#ff7300"><path d="m108.77 92.377c1.852-5.596 3.596-10.535 3.874-11.25 1.814-4.72 8.661-8.034 14.468-8.034h34.725l10.882-30.449 100.25 0.003-10.985 30.47h37.937s-2.829 8.661-6.361 19.26h23.762c10.123-28.101 18.159-50.359 18.159-50.359l-36.82-27.387h-23.006-40.46-70.211s-45.854-1.201-61.404 39.854c-4.59 12.082-9.365 24.939-14.106 37.893h19.285z" clip-path="url(#a)"/><path d="m278.63 158.96c-0.299 0.893-1.898 5.685-1.994 5.973-3.601 10.717-15.417 10.23-13.861 10.23h-37.864l-10.648 31.476h-100.5l11.311-31.321h-49.078l4.819-15.746h-15.614l-17.008 48.849 38.671 26.786h23.003 87.635 23.034s46.504-1.679 57.705-33.796c2.575-7.377 7.758-23.68 14.5-42.45h-14.112z" clip-path="url(#a)"/><polygon clip-path="url(#a)" points="114.87 106.73 107.04 110.89 106.78 111.03 107.12 111.03 130.39 111.03 130.6 111.03 130.75 110.96 140.42 106.76 127.66 141.11 116.74 147.02 136.73 147.02 149.63 111.03 172.33 111.03 175.48 102.49 175.45 102.49 167.63 106.73"/><polygon clip-path="url(#a)" points="235.48 106.73 228.13 110.63 227.38 111.03 228.14 111.03 250.68 111.03 251.2 111.03 251.36 110.96 260.99 106.77 248.27 141.11 237.33 147.02 257.33 147.02 270.23 111.03 292.94 111.03 296.09 102.49 296.07 102.49 288.24 106.73"/><path d="m211.35 132.31h-34.329l12.842-4.315h-0.208l14.298-18.04-0.007 0.151 7.751-4.602-0.037 0.01h-3.608c-10.021 0.268-18.681 10.831-18.681 10.831l-22.055 25.068-16.392 5.578-0.026 0.024h23.717l8.353-10.528h19.55l8.832-4.177z" clip-path="url(#a)"/><polygon clip-path="url(#a)" points="211.83 105.46 211.83 141.79 201.84 147 222.49 147 221.49 111.76 221.35 102.5"/><path d="m56.394 111.1h45.87l2.928-8.563-8.004 4.209h-40.105c-5.395 0-8.077 2.271-9.131 6.028-0.731 2.605-1.842 7.185-1.842 7.185l-6.996 3.786c2.019 1.494 4.722 2.334 7.309 2.959 5.494 1.339 28.558 7.357 28.772 7.412l10.432-4.178s-1.873 5.057-2.587 6.835c-0.573 1.435-3.108 5.901-10.032 5.901h-41.079l-7.966 4.294-0.003 0.014s33.985 0.007 50.071 0.007c16.462 0 18.422-8.28 19.358-11.122 0.63-1.916 1.248-5.747-1.537-8.901l-6.239 2.959-39.493-9.976 8.647-4.418 1.627-4.431z" clip-path="url(#a)"/><path d="m360.74 102.51-7.896 4.207h-35.454c-9.227 0-11.324 6.754-12.547 10.179-1.219 3.416-9.021 24.709-9.021 24.709l-10.212 5.403 59.168 0.035 3.103-8.585h-0.014l-5.85 4.329h-42.36l8.283-4.367 3.557-9.478h22.165l3.495-9.771-5.979 3.201h-22.121l6.048-3.201 2.822-8.143h39.807l3.028-8.517h-0.022z" clip-path="url(#a)"/></g><g fill="#fff"><path d="m126.52 89.754c-2.64 0-3.431-0.192-2.771-2.197l1.111-3.019c0.805-1.979 1.74-2.176 4.384-2.176 2.636 0 3.501 0.197 2.773 2.176l-1.11 3.019c-0.739 2.005-1.747 2.197-4.387 2.197m-1.143 3.112c5.153 0 9.097-0.788 10.616-4.911l1.392-3.793c1.519-4.123-1.845-4.911-7.003-4.911-5.152 0-9.097 0.788-10.612 4.911l-1.396 3.793c-1.517 4.123 1.851 4.911 7.003 4.911" clip-path="url(#a)"/><polygon points="154.6 79.36 160.74 79.36 151.62 85.93 156.37 92.763 150.08 92.763 146.71 87.507 145.18 87.507 143.25 92.763 138.17 92.763 143.1 79.36 148.18 79.36 146.33 84.401 147.85 84.401" clip-path="url(#a)"/><polygon points="171.93 92.763 158.9 92.763 163.83 79.358 168.91 79.358 165.12 89.651 173.07 89.651" clip-path="url(#a)"/><path d="m188.09 87.846h-4.717l4.442-5.803h0.05l0.225 5.803zm0.057 2.609 0.145 2.305h5.178l-0.857-13.403h-7.721l-10.641 13.403h5.279l1.79-2.305h6.827z" clip-path="url(#a)"/><polygon points="210.21 87.506 203.59 87.506 201.66 92.761 196.58 92.761 201.51 79.358 206.59 79.358 204.73 84.399 211.36 84.399 213.21 79.358 218.29 79.358 213.36 92.761 208.28 92.761" clip-path="url(#a)"/><path d="m228.25 89.754c-2.644 0-3.434-0.192-2.774-2.197l1.112-3.019c0.804-1.979 1.744-2.176 4.384-2.176 2.637 0 3.505 0.197 2.777 2.176l-1.112 3.019c-0.728 2.005-1.744 2.197-4.387 2.197m-1.146 3.112c5.155 0 9.097-0.788 10.621-4.911l1.394-3.793c1.514-4.123-1.851-4.911-7-4.911-5.159 0-9.1 0.788-10.624 4.911l-1.394 3.793c-1.518 4.123 1.853 4.911 7.003 4.911" clip-path="url(#a)"/><polygon points="245.81 79.358 253.85 79.358 254.38 87.777 254.56 87.777 261.11 79.358 269.36 79.358 264.43 92.761 259.36 92.761 263.64 82.483 263.46 82.483 254.87 92.761 250.47 92.761 249.38 82.483 249.22 82.483 245.95 92.761 240.88 92.761" clip-path="url(#a)"/><path d="m282.03 87.846h-4.717l4.442-5.803h0.051l0.224 5.803zm0.059 2.609 0.144 2.305h5.181l-0.865-13.403h-7.714l-10.643 13.403h5.279l1.789-2.305h6.829z" clip-path="url(#a)"/><path d="m106.95 155.79-3.203 8.706c-1.279 3.481-4.06 4.456-9.727 4.456-7.726 0-8.063-1.322-6.914-4.456l3.203-8.706h5.156l-3.181 8.637c-0.542 1.476 0.685 1.493 2.851 1.493 2.19 0 2.983-0.113 3.546-1.651l3.115-8.479h5.154z" clip-path="url(#a)"/><polygon points="112.98 155.79 121.69 155.79 123.54 165.82 123.75 165.82 127.12 155.79 132.28 155.79 127.48 168.85 118.69 168.85 116.92 158.82 116.71 158.82 113.33 168.85 108.18 168.85" clip-path="url(#a)"/><polygon points="138.23 155.79 143.38 155.79 138.58 168.85 133.43 168.85" clip-path="url(#a)"/><polygon points="156.2 168.85 148.17 168.85 147.51 155.79 152.74 155.79 153.21 165.86 153.37 165.86 161.32 155.79 166.65 155.79" clip-path="url(#a)"/><polygon points="180.9 163.46 172.65 163.46 171.72 166 180.68 166 179.63 168.85 165.52 168.85 170.32 155.79 184.25 155.79 183.2 158.65 174.42 158.65 173.54 161.03 181.79 161.03" clip-path="url(#a)"/><path d="m192.28 162.35 1.294-3.529h3.992c1.706 0 2.012 0.326 1.593 1.493-0.546 1.483-0.676 2.036-2.885 2.036h-3.994m-7.546 6.499h5.156l1.276-3.467h3.989c1.547 0 1.921 0.666 1.509 1.638l-0.671 1.83h5.147l0.971-2.637c0.543-1.476-0.524-2.259-2.626-2.31l0.049-0.137c3.872-0.577 4.339-1.847 5.142-4.02 1.202-3.254-0.755-3.896-5.193-3.955h-9.943l-4.806 13.058z" clip-path="url(#a)"/><path d="m217.79 160.92c5.623 0.278 6.859 0.762 5.761 3.756-0.797 2.163-1.448 4.284-9.948 4.284-4.923 0-8.163-0.069-6.591-4.356h4.943c-0.498 1.363 0.175 1.5 2.698 1.5 2.416 0 3.254-0.103 3.652-1.198 0.422-1.129 0.13-1.116-2.245-1.215l-1.119-0.058c-4.679-0.23-7.477-0.209-6.097-3.948 1.38-3.745 4.586-3.989 9.636-3.989 4.223 0 7.816 0.055 6.484 3.649l-0.192 0.549h-4.95c0.495-1.342-0.333-1.342-2.393-1.342-2.776 0-3.154 0.385-3.422 1.112-0.385 1.044 0.35 1.082 1.688 1.15l2.095 0.106z" clip-path="url(#a)"/><polygon points="231.72 155.79 236.86 155.79 232.06 168.85 226.92 168.85" clip-path="url(#a)"/><polygon points="243.93 158.82 238.98 158.82 240.1 155.79 255.4 155.79 254.28 158.82 249.08 158.82 245.4 168.85 240.24 168.85" clip-path="url(#a)"/><polygon points="257.93 168.85 259.61 164.27 255.88 155.79 261.77 155.79 263.37 160.83 263.56 160.83 268.95 155.79 274.93 155.79 264.76 164.27 263.07 168.85" clip-path="url(#a)"/></g><g stroke="#f90" fill="#f90"><polygon clip-path="url(#a)" points="65.201 159.57 48.228 208.34 82.356 189.8 87.459 175.29 75.944 175.29 80.891 159.57" stroke-width=".8"/><path d="m108.77 92.377c1.852-5.596 3.181-9.332 3.385-10.067 1.178-4.25 6.768-7.962 11.692-8.884l6.33-17.477s9.709-27.694 37.402-27.694l133.23 0.026-2.152-13.651h-136.34s-44.443 0.106-59.856 41.227l-13.118 36.52h19.428z" clip-path="url(#a)" stroke-width="0.8"/><g stroke-width=".8"><path d="m224.86 175.3-10.566 31.088-100.52 0.24-27.011 14.185h124.04s27.02-0.587 33.508-20.191l7.792-25.314-27.243-0.008z" clip-path="url(#a)"/><polygon clip-path="url(#a)" points="272.96 42.647 262.14 73.117 286.23 73.117 300.8 28.281"/><path d="m105.24 102.51h-50.541c-4.222 0-14.359 1.445-17.952 9.912-3.455 8.153 2.348 11.33 2.348 11.33l7.017-3.797s1.111-4.579 1.845-7.185c1.054-3.761 3.737-6.032 9.128-6.032h40.105l8.05-4.228z" clip-path="url(#a)"/><path d="m54.784 115.52-8.664 4.43 39.493 9.972 6.253-2.961c-1.617-1.883-4.116-3.11-8.235-4.186-1.565-0.429-1.653-0.403-28.847-7.255" clip-path="url(#a)"/><path d="m27.251 138.47-3.308 8.506 7.983-4.308h41.083c6.92 0 9.459-4.466 10.029-5.901 0.716-1.775 2.586-6.835 2.586-6.835l-10.436 4.181-1.505 4.356h-46.432z" clip-path="url(#a)"/><polygon clip-path="url(#a)" points="167.63 106.73 175.48 102.47 109.88 102.47 106.78 111.03 114.87 106.73"/><polygon clip-path="url(#a)" points="140.42 106.76 130.6 111.03 116.71 147.04 127.66 141.11"/><path d="m221.38 102.49h-16.573s-12.734-0.651-25.333 14.282l-20.557 22.286-8.01 7.933 16.409-5.582 22.054-25.07s8.664-10.56 18.678-10.827h3.611l9.721-3.022z" clip-path="url(#a)"/><polygon clip-path="url(#a)" points="189.91 127.97 177.02 132.31 211.35 132.31 202.52 136.48 201.84 147 211.84 141.79 211.83 105.42 203.91 110.12 203.49 127.97"/><polygon clip-path="url(#a)" points="288.23 106.73 296.13 102.46 230.46 102.49 227.38 111.03 235.48 106.73"/><polygon clip-path="url(#a)" points="260.99 106.77 251.2 111.03 237.29 147.04 248.26 141.11"/><path d="m360.78 102.49-42.581 0.053s-15.221-1.21-20.683 12.169l-9.234 24.469-2.701 7.844 10.243-5.417s7.799-21.297 9.015-24.712c1.222-3.421 3.326-10.173 12.55-10.173h35.454l7.937-4.233z" clip-path="url(#a)"/><polygon clip-path="url(#a)" points="309.05 122.38 331.18 122.38 337.15 119.17 315.1 119.17"/><polygon clip-path="url(#a)" points="299.66 142.79 342.02 142.79 347.91 138.43 307.96 138.41"/></g></g>
</svg>
</a>

    </div>
    
        <div class="searchbox">
    <label for="search-by"><i class="fa fa-search"></i></label>
    <input data-search-input id="search-by" type="text" placeholder="Search...">
    <span data-search-clear=""><i class="fa fa-close"></i></span>
</div>

<script type="text/javascript" src="/js/lunr.min.js?1521515504"></script>
<script type="text/javascript" src="/js/auto-complete.js?1521515504"></script>
<script type="text/javascript">
    
        var baseurl = '\/';
    
</script>
<script type="text/javascript" src="/js/search.js?1521515504"></script>

    
  </div>

    <div class="highlightable">
    <ul class="topics">

        
          
          


 
  
    
    <li data-nav-id="/start/" title="Getting Started" class="dd-item 
        
        
        
        ">
      <a href="/start/">
          <b>1. </b>Getting Started
          
            <i class="fa fa-check read-icon"></i>
          
      </a>
      
              
    </li>
  
 

          
          


 
  
    
    <li data-nav-id="/morning/" title="Morning Session" class="dd-item 
        
        
        
        ">
      <a href="/morning/">
          <b>2. </b>Morning Session
          
            <i class="fa fa-check read-icon"></i>
          
      </a>
      
      
        <ul>
          
          
          
          
        
          
            
            


 
  
    
      <li data-nav-id="/morning/links/" title="Resources for new users" class="dd-item ">
        <a href="/morning/links/">
        <b>2.1 </b>Resources for new users
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/morning/cowboy/" title="Introduction to the Cowboy" class="dd-item ">
        <a href="/morning/cowboy/">
        <b>2.2 </b>Introduction to the Cowboy
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/morning/basic_unix/" title="Unix overview" class="dd-item ">
        <a href="/morning/basic_unix/">
        <b>2.3 </b>Unix overview
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/morning/unix_essentials/" title="Unix essentials" class="dd-item ">
        <a href="/morning/unix_essentials/">
        <b>2.4 </b>Unix essentials
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/morning/answers/" title="Answers!" class="dd-item ">
        <a href="/morning/answers/">
        <b>2.5 </b>Answers!
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
        
        </ul>
              
    </li>
  
 

          
          


 
  
    
    <li data-nav-id="/afternoon/" title="Afternoon Session" class="dd-item 
        parent
        
        
        ">
      <a href="/afternoon/">
          <b>3. </b>Afternoon Session
          
            <i class="fa fa-check read-icon"></i>
          
      </a>
      
      
        <ul>
          
          
          
          
        
          
            
            


 
  
    
      <li data-nav-id="/afternoon/terminology/" title="Terminology" class="dd-item ">
        <a href="/afternoon/terminology/">
        <b>3.1 </b>Terminology
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/afternoon/data_acquisition/" title="Data Acquisition" class="dd-item active">
        <a href="/afternoon/data_acquisition/">
        <b>3.2 </b>Data Acquisition
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/afternoon/regex/" title="Regular expressions" class="dd-item ">
        <a href="/afternoon/regex/">
        <b>3.3 </b>Regular expressions
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/afternoon/my_first_script/" title="My first bash script" class="dd-item ">
        <a href="/afternoon/my_first_script/">
        <b>3.4 </b>My first bash script
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/afternoon/versions_git/" title="Version with with git" class="dd-item ">
        <a href="/afternoon/versions_git/">
        <b>3.5 </b>Version with with git
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
            
            


 
  
    
      <li data-nav-id="/afternoon/sources/" title="Sources afternoon" class="dd-item ">
        <a href="/afternoon/sources/">
        <b>3.6 </b>Sources afternoon
        <i class="fa fa-check read-icon"></i>
        </a>
    </li>
     
  
 

            
          
        
        </ul>
              
    </li>
  
 

          
          


 
  
    
    <li data-nav-id="/wrap-up/" title="Wrap-up" class="dd-item 
        
        
        
        ">
      <a href="/wrap-up/">
          <b>4. </b>Wrap-up
          
            <i class="fa fa-check read-icon"></i>
          
      </a>
      
              
    </li>
  
 

          
         
    </ul>

    
    
      <section id="shortcuts">
        <h3>More</h3>
        <ul>
          
              <li> 
                  <a class="padding" href="https://github.com/andrese52/epp_workshop_web"><i class='fa fa-fw fa-github'></i> Github repo</a>
              </li>
          
              <li> 
                  <a class="padding" href="https://bioinformatics.netlify.com/credits"><i class='fa fa-fw fa-bullhorn'></i> Credits</a>
              </li>
          
        </ul>
      </section>
    

    
    <section id="prefooter">
      <hr/>
      <ul>
       
      
      
        <li><a class="padding" href="#" data-clear-history-toggle=""><i class="fa fa-fw fa-history"></i> Clear History</a></li>         
      
      </ul>
    </section>
    
    <section id="footer">
      <center>
    
    <a class="github-button" href="https://github.com/andrese52/epp_workshop_web/archive/master.zip" data-icon="octicon-cloud-download" aria-label="Download this website on GitHub">Download</a>

    
    <a class="github-button" href="https://github.com/andrese52/epp_workshop_web" data-icon="octicon-star" data-show-count="true" aria-label="Star this website on GitHub">Star</a>

    
    <a class="github-button" href="https://github.com/andrese52/epp_workshop_web/fork" data-icon="octicon-repo-forked" data-show-count="true" aria-label="Fork this website on GitHub">Fork</a>

    <p>Copyright 2018</p>
	<p> <a href="https://www.aespindola.com"><i class="fa fa-adn fa-lg"></i></a> <a href="https://www.aespindola.com"></a></p>
</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>





        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
              
              <div>
                <div id="top-bar">
                
                  
                  
                  
                  <div id="top-github-link">
                    <a class="github-link" title='Edit this page' href="https://bioinformatics.netlify.com/afternoon/data_acquisition.md" target="blank">
                      <i class="fa fa-code-fork"></i>
                      <span id="top-github-link-text">Edit this page</span>
                    </a>
                  </div>
                  
                
                
                <div id="breadcrumbs" itemscope="" itemtype="http://data-vocabulary.org/Breadcrumb">
                    <span id="sidebar-toggle-span">
                        <a href="#" id="sidebar-toggle" data-sidebar-toggle="">
                          <i class="fa fa-bars"></i>
                        </a>
                    </span>
                  
                  <span id="toc-menu"><i class="fa fa-list-alt"></i></span>
                  
                  <span class="links">
                    
          
          
            
            
          
          
            
            
          
          
            <a href='/'>Introduction to Programming for beginners</a> > <a href='/afternoon/'>Afternoon Session</a> > Data Acquisition
          
         
          
         
          
           
                  </span>
                </div>
                
                    <div class="progress">
    <div class="wrapper">
<nav id="TableOfContents">
<ul>
<li>
<ul>
<li><a href="#contents">Contents</a></li>
<li><a href="#intro">Intro</a></li>
<li><a href="#back-to-the-terminal">Back to the terminal</a></li>
<li><a href="#retrieving-my-first-ftp-file">Retrieving my first ftp file</a></li>
<li><a href="#filtering-data">Filtering Data</a>
<ul>
<li><a href="#loops-and-variables-in-linux">Loops and variables in  Linux</a>
<ul>
<li><a href="#task-1">TASK 1:</a></li>
<li><a href="#solution-1">Solution 1</a></li>
</ul></li>
<li><a href="#file-compression">File compression</a>
<ul>
<li><a href="#task-2">TASK 2</a></li>
<li><a href="#solution-2">Solution 2</a></li>
</ul></li>
</ul></li>
</ul></li>
</ul>
</nav>
    </div>
</div>

                
              </div>
            </div>
            

        
        <div id="body-inner">
          
            <h1>Data Acquisition</h1>
          

        




<h2 id="contents">Contents</h2>

<ul>
<li><a href="#Intro">Introduction</a></li>
<li><a href="#back-to-the-terminal">Back to the terminal</a></li>
<li><a href="#retrieving-my-first-ftp-file">Retrieving my first ftp file</a></li>
<li><a href="#filtering-data">Filtering Data</a></li>
<li><a href="#loops-and-variables-in-linux">Loops and variables in  Linux</a></li>
<li><a href="#task-1">Task 1</a></li>
<li><a href="#task-2">Task 2</a></li>
</ul>

<h2 id="intro">Intro</h2>

<p>[<a href="#contents">back to top</a>]</p>

<div class="notices note" ><p>While using coding and bioinformatics, we want to
avoid using graphical user interphases (GUIs) as much as we can. Can
somebody answer why?</p>
</div>


<p>Visualizing large datasets (as the ones that we are going to download)
could be practically impossible.  Opening the whole document will likely
<strong>crash your computer</strong>.</p>


<figure>
    
        <img src="https://media.giphy.com/media/l3q2Mdm0cRPLPcuGI/giphy.gif" />
    
    
</figure>


<p>However, there are exceptions, where we already know the data structure
of a file and we can start running codes on it.</p>

<p><a href="https://www.ncbi.nlm.nih.gov/">NCBI</a> which stands for the The National
Center for Biotechnology Information helps bioinformaticians by
providing access to biomedical and genomic information. This is a GUI
which helps people search and process data. This is given the fact that
the data has been previously curated by others.</p>

<p>They also have access to an <a href="ftp://ftp.ncbi.nlm.nih.gov">FTP site</a>,
where we can download files in bulk. Usually, exploring these folders
become difficult, therefore, they create summary files that we can play
with.</p>

<p>Let&rsquo;s start by downloading the <code>assembly_summary_genbank.txt</code> from the
NCBI FTP site at <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/">ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/</a>.
If you click on this site link the directory content should look
something like this:</p>

<pre><code>.
├── [ 30] README_assembly_summary.txt -&gt; ../README_assembly_summary.txt
├── [ 27] README_change_notice.txt -&gt; ../README_change_notice.txt
├── [ 22K] species_genome_size.txt.gz
├── [ 31M] ANI_report_bacteria.txt
├── [2.2M] assembly_summary_genbank_historical.txt
├── [ 46M] assembly_summary_genbank.txt
├── [2.4M] assembly_summary_refseq_historical.txt
└── [ 34M] assembly_summary_refseq.txt
</code></pre>

<h2 id="back-to-the-terminal">Back to the terminal</h2>

<p>[<a href="#contents">back to top</a>]</p>

<p>Create a new folder in your <code>/scratch/username</code> directory named
<code>afternoon</code> and move inside the <code>afternoon</code> folder:</p>

<pre><code class="language-bash">$ mkdir afternoon 
$ cd afternoon 
</code></pre>

<p>Let&rsquo;s make sure that you are in the correct directory by typing <code>pwd</code></p>

<pre><code class="language-bash">$ pwd 
</code></pre>

<p>you should get something like this:</p>

<pre><code>/scratch/asecas86/afternoon
</code></pre>

<blockquote>
<p>My username is <strong><em>asecas86</em></strong> (you already learned this)</p>
</blockquote>

<p>Now, let us retrieve the <code>assembly_summary_genbank.txt</code></p>

<h2 id="retrieving-my-first-ftp-file">Retrieving my first ftp file</h2>

<p>[<a href="#contents">back to top</a>]</p>

<pre><code class="language-bash">$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt 
</code></pre>

<pre><code>--2018-03-04 22:46:48-- 
ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt
                   =&gt; “assembly_summary_genbank.txt”
        Resolving ftp.ncbi.nlm.nih.gov... 165.112.9.229, 
2607:f220:41e:250::7
        Connecting to ftp.ncbi.nlm.nih.gov|165.112.9.229|:21... 
connected.
        Logging in as anonymous ... Logged in!
        ==&gt; SYST ... done.  ==&gt; PWD ... done.
        ==&gt; TYPE I ... done.  ==&gt; CWD (1) 
/genomes/ASSEMBLY_REPORTS ... done.
        ==&gt; SIZE assembly_summary_genbank.txt ... 48113668
        ==&gt; PASV ... done.  ==&gt; RETR 
assembly_summary_genbank.txt ... done.
        Length: 48113668 (46M) (unauthoritative)

100%[=====================================================================================================================================&gt;] 
48,113,668 11.1M/s in 8.4s 
</code></pre>

<p>Let&rsquo;s verify that that the file is present</p>

<pre><code class="language-bash">$ ls -lth 
</code></pre>

<pre><code>total 52M
-rw------- 1 asecas86 clusterusers 46M Mar 4 22:46 assembly_summary_genbank.txt 
</code></pre>

<p>Let&rsquo;s explore the head of the file</p>

<pre><code class="language-bash">$ head -n 2 assembly_summary_genbank.txt 
</code></pre>

<pre><code>#   See ftp://ftp.ncbi.nlm.nih.gov/genomes/README_assembly_summary.txt for a description of the columns in this file.
# assembly_accession    bioproject      biosample       wgs_master      refseq_category taxid   species_taxid   organism_name   infraspecific_name      isolate       version_status  assembly_level  release_type    genome_rep      seq_rel_date    asm_name        submitter       gbrs_paired_asm paired_asm_comp ftp_path      excluded_from_refseq    relation_to_type_material
GCA_000001215.4 PRJNA13812      SAMN02803731            reference genome        7227    7227    Drosophila melanogaster                 latest  Chromosome   Major    Full    2014/08/01      Release 6 plus ISO1 MT  The FlyBase Consortium/Berkeley Drosophila Genome Project/Celera Genomics       GCF_000001215.4 identical     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/215/GCA_000001215.4_Release_6_plus_ISO1_MT
</code></pre>

<p>Can we make sense of the data from here?  Really difficult to tell
right??? That&rsquo;s why it is important to <strong>ALWAYS CREATE A README FILE</strong>.</p>

<p>Let&rsquo;s take a peek at their <code>README_assembly_summary.txt</code>.</p>

<pre><code class="language-bash">$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/README_assembly_summary.txt 
</code></pre>

<p><code>nano</code> is utilized to edit files. However, we just want to read the file so let&rsquo;s use <code>less</code></p>

<pre><code class="language-bash">$ less README_assembly_summary.txt
</code></pre>

<p>Keep pushing <code>enter</code> until you reach the end of the document and tell me what information about their ftp archives can be recovered and are useful. Take notes of what information we can get on each column.</p>

<div class="notices warning" >Push <code>Q</code> to quit watching the <code>readme</code> file.</div>


<h2 id="filtering-data">Filtering Data</h2>

<p>[<a href="#contents">back to top</a>]</p>

<p>Let&rsquo;s focus on our organism of interest <code>Aspergillus flavus</code>. Let&rsquo;s try searching for our pathogen in the database:</p>

<pre><code class="language-bash">$ grep &quot;aspergillus flavus&quot; assembly_summary_genbank.txt
</code></pre>

<p><strong><em>What did you get as output from the previous code?</em></strong> <strong>NOTHING? HOW COME??</strong></p>

<p>Try this</p>

<pre><code class="language-bash">$ grep -i &quot;aspergillus flavus&quot; assembly_summary_genbank.txt
</code></pre>

<div class="notices note" >Take 1 minute to understand why this happened</div>


<p>You are right, this is because you added the <code>-i</code> argument which asks for <code>case insensitive</code></p>

<ul>
<li>Let&rsquo;s count how many entries for <code>Aspergillus flavus</code> we have and at the same time let&rsquo;s redirect our entries to a file named <code>aspergillus_assemblies.txt</code></li>
</ul>

<pre><code class="language-bash">$ grep -ic &quot;aspergillus flavus&quot; assembly_summary_genbank.txt
$ grep -i &quot;aspergillus flavus&quot; assembly_summary_genbank.txt &gt; aspergillus_assemblies.txt 
</code></pre>

<pre><code>17
</code></pre>

<p>We do have 17 assemblies of <code>Aspergillus flavus</code> available in this FTP site. Guess what, not all of them are of importance to us. I want only assemblies at the chromosome level. Remember that there can be different types of assembly levels <a href="../../afternoon/terminology/#sequence-assembly">See above</a>. We want to sort by column to select those Chromosomal assemblies (if any). Now that we have retrieved data we will learn how to filter by column in the next session.</p>

<p>Well, let me introduce you to the beautiful <code>awk</code> (We will be seeing more of <code>awk</code> under <a href="../regex">Regular Expressions</a>)</p>

<pre><code class="language-bash">$ awk -F'\t' '$12 == &quot;Chromosome&quot;' aspergillus_assemblies.txt | wc -l 
</code></pre>

<pre><code>    0
</code></pre>

<p><strong>Can somebody tell me what are we doing with the above code?</strong></p>

<p>We are piping the output of <code>awk</code> to a <code>wc</code> command that is the famous <code>word count</code>. You can count how many lines are found in a file with <code>wc -l</code>.</p>

<p>Unfortunately, we have <code>0</code> crhomosomal assemblies. Now let&rsquo;s try Scaffolds.</p>

<h3 id="loops-and-variables-in-linux">Loops and variables in  Linux</h3>

<p>[<a href="#contents">back to top</a>]</p>

<p>Assigning values to variables is really easy, just type the following:</p>

<pre><code class="language-bash">a=&quot;Andres&quot;
</code></pre>

<p>Now if you want to retrieve your variable and print it to the prompt do the following.</p>

<pre><code class="language-bash">echo $a
</code></pre>

<p>For the next task you need to know how loops work in linux.</p>

<p>Let&rsquo;s try some, but first let&rsquo;s get some variables assigned</p>

<pre><code class="language-bash">a=&quot;like&quot;
b=&quot;this&quot;
c=&quot;workshop&quot;
</code></pre>

<pre><code class="language-bash">for i in $a $b $c
do
echo $i
done
</code></pre>

<blockquote>
<h4 id="task-1">TASK 1:</h4>

<p>[<a href="#contents">back to top</a>]</p>

<p>Complete the following tab-delimited table with information about how many assemblies are at the chromosome, scaffold and contig level. The table must look something like this:</p>

<table>
<thead>
<tr>
<th>Assembly level</th>
<th>Number of assemblies</th>
</tr>
</thead>

<tbody>
<tr>
<td>Chromosome</td>
<td>0</td>
</tr>

<tr>
<td>Scaffold</td>
<td>&ndash;</td>
</tr>

<tr>
<td>Contig</td>
<td>&ndash;</td>
</tr>
</tbody>
</table>

<div class="notices tip" ><p><strong>HINT:</strong> You can create a for loop</p>
</div>

</blockquote>

<h4 id="solution-1">Solution 1</h4>

<p>First we need to know which list will be used in our for loop. In this case I want a list of <code>Chromosome, Scaffold and Contig</code>. Then you create the <code>loop</code> like so:</p>

<pre><code class="language-bash">for i in Chromosome Scaffold Contig
do
assemblies=`awk -v var=&quot;$i&quot; -F'\t' '$12 == Chromosome' aspergillus_assemblies.txt | wc -l`
echo -e $i'\t'$assemblies &gt;&gt; table_assemblies.txt
echo -e $i'\t'$assemblies
done
</code></pre>

<p>Did you really try to only copy and paste the loop and got an error? Try harder&hellip;.. Focus on the <code>var</code></p>

<p>The result should look something like this:</p>

<pre><code>[asecas86@n252 afternoon]$ for i in Chromosome Scaffold Contig
&gt; do
&gt; assemblies=`awk -v var=&quot;$i&quot; -F'\t' '$12 == var' aspergillus_assemblies.txt | wc -l`
&gt; echo -e $i ' \t ' $assemblies &gt;&gt; table_assemblies.txt
&gt; echo -e $i ' \t ' $assemblies
&gt; done
Chromosome        0
Scaffold          5
Contig            12
</code></pre>

<p>Let us retrieve only the best assemblies, in this case lets select <code>Scaffolds</code></p>

<pre><code class="language-bash">$ awk -F'\t' '$12 == &quot;Scaffold&quot;' aspergillus_assemblies.txt &gt; chr_aspergillus_assemblies.txt
</code></pre>

<p>Let&rsquo;s check those 5 entries by using <code>head</code></p>

<pre><code class="language-bash">$ head chr_aspergillus_assemblies.txt
</code></pre>

<p>I want to get the ftp address to download those files.</p>

<pre><code class="language-bash">$ awk -F'\t' '{print $20}' chr_aspergillus_assemblies.txt &gt; ftp-ids.txt
</code></pre>

<p>and finally check if you got the <code>five</code> ftp addresses.</p>

<pre><code class="language-bash">$ less ftp-ids.txt
</code></pre>

<p>We want to compare these five genomes.  Therefore, we want to get the genome files and the annotation files (fasta and gff or gtf)</p>

<p>You can open one of the files to explore the directory, for example, let&rsquo;s explore this:  <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/443/195/GCA_002443195.1_AflaGuard">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/443/195/GCA_002443195.1_AflaGuard</a></p>


<figure>
    
        <img src="../../images/aflaguard.png" />
    
    
</figure>


<p>There are multiple files inside this directory, we are interested in <code>.fna</code> and <code>gff</code> or <code>gtf</code> files. Let&rsquo;s target those files with a <code>wildcard</code> and a <code>while</code> loop.</p>

<p>First let&rsquo;s make a dir for the genomes and annotations that we will retrieve:</p>

<pre><code class="language-bash">$ mkdir genomes
</code></pre>

<p>Now finally let us retrieve those files with the following loop:</p>

<p><strong>Very important note:</strong> Make sure that the space between <code>wget -P genome/</code> and <code>$p/*.fna.gz</code> is present, otherwise <code>wget</code> will not work. Can you guess why??</p>

<pre><code class="language-bash">while read p; do
  echo $p
wget -P genomes/ $p/*.fna.gz
done &lt;ftp-ids.txt
</code></pre>

<p>Now check what files you downloaded:</p>

<pre><code class="language-bash">$ ls -lth genomes/
</code></pre>

<pre><code>total 77M
-r--r--r-- 1 asecas86 clusterusers  12M Jan  5 22:21 GCA_002864195.1_ASM286419v1_genomic.fna.gz
-r--r--r-- 1 asecas86 clusterusers  11M Dec 14 01:09 GCA_002456175.1_AF36_genomic.fna.gz
-r--r--r-- 1 asecas86 clusterusers  12M Nov 17 11:31 GCA_002443215.1_K49_genomic.fna.gz
-r--r--r-- 1 asecas86 clusterusers  11M Nov 17 11:29 GCA_002443195.1_AflaGuard_genomic.fna.gz
-r--r--r-- 1 asecas86 clusterusers 5.9M Nov 13 05:26 GCA_000006275.2_JCVI-afl1-v2.0_cds_from_genomic.fna.gz
-r--r--r-- 1 asecas86 clusterusers 5.9M Nov 13 05:26 GCA_000006275.2_JCVI-afl1-v2.0_rna_from_genomic.fna.gz
-r--r--r-- 1 asecas86 clusterusers  12M Jun 16  2016 GCA_000006275.2_JCVI-afl1-v2.0_genomic.fna.gz
</code></pre>

<p>you should have 5 files, why do we have 7?</p>

<p>This is because these folders contain all information about annotations and sometimes they include <code>CDS</code> which stand for coding regions or <code>rna</code> which are regions that were mapped to the genome using its transcripts. It seems like the only one that has this issue is <code>JCVI</code> assembly.</p>

<h3 id="file-compression">File compression</h3>

<p>Most files in bioinformatics are compressed to allow fast sharing, otherwise they can weight many Gigabytes. The most common compression that you will find online is <code>.gz</code>.  The best way to uncompress this type of files is by typing <code>gunzip *.gz</code></p>

<p>We can try that with a loop</p>

<pre><code class="language-bash">$ cd genomes
</code></pre>

<pre><code class="language-bash">for i in *.gz
do
echo &quot;Unzipping $i&quot;
gunzip $i
done 
</code></pre>

<p>DONE, now you have <code>fasta</code> files to analyze. And this is the way you retrieve data in bulk from ftp sites.
Notice that the directory size increased to 247 MB.</p>

<pre><code>total 247M
-r--r--r-- 1 asecas86 clusterusers 37M Jan  5 22:21 GCA_002864195.1_ASM286419v1_genomic.fna
-r--r--r-- 1 asecas86 clusterusers 36M Dec 14 01:09 GCA_002456175.1_AF36_genomic.fna
-r--r--r-- 1 asecas86 clusterusers 36M Nov 17 11:31 GCA_002443215.1_K49_genomic.fna
-r--r--r-- 1 asecas86 clusterusers 36M Nov 17 11:29 GCA_002443195.1_AflaGuard_genomic.fna
-r--r--r-- 1 asecas86 clusterusers 20M Nov 13 05:26 GCA_000006275.2_JCVI-afl1-v2.0_cds_from_genomic.fna
-r--r--r-- 1 asecas86 clusterusers 20M Nov 13 05:26 GCA_000006275.2_JCVI-afl1-v2.0_rna_from_genomic.fna
-r--r--r-- 1 asecas86 clusterusers 36M Jun 16  2016 GCA_000006275.2_JCVI-afl1-v2.0_genomic.fna
</code></pre>

<blockquote>
<h4 id="task-2">TASK 2</h4>

<p>[<a href="#contents">back to top</a>]</p>

<p>Download all the <code>gff</code> or <code>gtf</code> files associated to these five genomes and put them in a folder called <code>annotations</code>
<div class="notices tip" ><p><strong>HINT:</strong> use <code>mkdir annotations</code> in your <code>afternoon</code> directory</p>
</div>
</p>
</blockquote>

<h4 id="solution-2">Solution 2</h4>

<pre><code class="language-bash">$ mkdir ../annotations
</code></pre>

<p>What would you change in the following code to get gff files</p>

<p><strong>Very important note:</strong> Make sure that the space between <code>wget -P annotations/</code> and <code>$p/*.gff.gz</code> is present, otherwise <code>wget</code> will not work. Can you guess why??</p>

<pre><code class="language-bash">while read p; do
	echo $p
	wget -P annotations/ $p/*.fna.gz
done &lt;ftp-ids.txt
</code></pre>

<p>You will not get a single cup of coffee if you do not solve this loop as expected.
 
<figure>
    
        <img src="https://media.giphy.com/media/7qV3yswT0K8hi/giphy.gif" />
    
    
</figure>
</p>

<p>What about this Solution?? (Think twice before copying and pasting this one):</p>

<pre><code class="language-bash">p=&quot;u r 2 Lazy&quot;
for i in {1..100}
do
	echo $p $i &quot;Times&quot;
done
</code></pre>

<p>Ok try the loop with <code>gff</code> but first:</p>

<pre><code class="language-bash">$ cd ..
</code></pre>

<p>Then:
<strong>Very important note:</strong> Make sure that the space between <code>wget -P annotations/</code> and <code>$p/*.gff.gz</code> is present, otherwise <code>wget</code> will not work. Can you guess why??</p>

<pre><code class="language-bash">while read p; do
	echo $p
	wget -P annotations/ $p/*.gff.gz
done &lt;ftp-ids.txt
</code></pre>

<p>Let&rsquo;s see what we got:</p>

<pre><code class="language-bash">$ ls -lth annotations/
</code></pre>

<pre><code>total 3.0M
-r--r--r-- 1 asecas86 clusterusers 4.2K Jan  5 22:21 GCA_002864195.1_ASM286419v1_genomic.gff.gz
-r--r--r-- 1 asecas86 clusterusers 3.3K Dec 14 01:09 GCA_002456175.1_AF36_genomic.gff.gz
-r--r--r-- 1 asecas86 clusterusers 3.3K Nov 17 11:31 GCA_002443215.1_K49_genomic.gff.gz
-r--r--r-- 1 asecas86 clusterusers 3.1K Nov 17 11:29 GCA_002443195.1_AflaGuard_genomic.gff.gz
-r--r--r-- 1 asecas86 clusterusers 2.3M Nov 13 05:26 GCA_000006275.2_JCVI-afl1-v2.0_genomic.gff.gz
</code></pre>

<p>Uncompress:</p>

<pre><code class="language-bash">$ cd annotations
</code></pre>

<p>and finally</p>

<pre><code class="language-bash">for i in *.gz
do
	echo &quot;Unzipping $i&quot;
	gunzip $i
done 
</code></pre>

<p>Check what you got:</p>

<pre><code class="language-bash">$ ls -lth
</code></pre>

<p>You should have 5 files. If you have more or less, then you did something wrong!</p>

<pre><code>total 29M
-r--r--r-- 1 asecas86 clusterusers 47K Jan  5 22:21 GCA_002864195.1_ASM286419v1_genomic.gff
-r--r--r-- 1 asecas86 clusterusers 31K Dec 14 01:09 GCA_002456175.1_AF36_genomic.gff
-r--r--r-- 1 asecas86 clusterusers 32K Nov 17 11:31 GCA_002443215.1_K49_genomic.gff
-r--r--r-- 1 asecas86 clusterusers 29K Nov 17 11:29 GCA_002443195.1_AflaGuard_genomic.gff
-r--r--r-- 1 asecas86 clusterusers 26M Nov 13 05:26 GCA_000006275.2_JCVI-afl1-v2.0_genomic.gff
</code></pre>

<p>In the upcoming modules you will learn how to iterate through the files that we just downloaded to compare basic metrics. Now, please get some fresh air.</p>

<p>[<a href="#contents">back to top</a>]</p>


<footer class=" footline" >
	
</footer>


        
        </div> 
        

      </div>

    <div id="navigation">
        
        
        
        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
        
        


        
            <a class="nav nav-prev" href="/afternoon/terminology/" title="Terminology"> <i class="fa fa-chevron-left"></i></a>
        
        
            <a class="nav nav-next" href="/afternoon/regex/" title="Regular expressions" style="margin-right: 0px;"><i class="fa fa-chevron-right"></i></a>
        
    </div>

    </section>
    
    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/js/clipboard.min.js?1521515504"></script>
    <script src="/js/perfect-scrollbar.min.js?1521515504"></script>
    <script src="/js/perfect-scrollbar.jquery.min.js?1521515504"></script>
    <script src="/js/jquery.sticky.js?1521515504"></script>
    <script src="/js/featherlight.min.js?1521515504"></script>
    <script src="/js/html5shiv-printshiv.min.js?1521515504"></script>
    <script src="/js/highlight.pack.js?1521515504"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/js/modernizr.custom.71422.js?1521515504"></script>
    <script src="/js/learn.js?1521515504"></script>
    <script src="/js/hugo-learn.js?1521515504"></script>

    <link href="/mermaid/mermaid.css?1521515504" type="text/css" rel="stylesheet" />
    <script src="/mermaid/mermaid.js?1521515504"></script>
    <script>
        mermaid.initialize({ startOnLoad: true });
    </script>
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-115379955-1', 'auto');
  ga('send', 'pageview');

</script>
  </body>
</html>

